crates/sui-core/src/execution_scheduler/execution_scheduler_impl.rs

```diff
@@ -26,19 +28,82 @@ use std::{
 };
 use sui_config::node::AuthorityOverloadConfig;
 use sui_types::{
-    base_types::{FullObjectID, SequenceNumber},
+    base_types::{FullObjectID, ObjectID, SequenceNumber},
     error::SuiResult,
     executable_transaction::VerifiedExecutableTransaction,
     storage::{ChildObjectResolver, InputKey},
-    transaction::{SenderSignedData, TransactionDataAPI, TransactionKey},
+    transaction::{
+        InputObjectKind, SenderSignedData, SharedObjectMutability, TransactionDataAPI,
+        TransactionKey,
+    },
     SUI_ACCUMULATOR_ROOT_OBJECT_ID,
 };
 use tokio::sync::mpsc::UnboundedSender;
+use tokio::sync::oneshot;
 use tokio::time::Instant;
-use tracing::{debug, error};
+use tracing::{debug, error, instrument};
 
 use super::{overload_tracker::OverloadTracker, PendingCertificate};
 
+enum ObjectWriteBarrier {
+    // Tracks notifies that are received before the waiter registers itself.
+    PendingWaiter {
+        count: u32,
+    },
+    // Tracks state after the waiter registers itself.
+    Waiting {
+        notify: Option<oneshot::Sender<()>>,
+        count: u32,   // how many notifies have been received
+        barrier: u32, // how many notifies are needed to pass the barrier
+    },
+}
+
+impl Default for ObjectWriteBarrier {
+    fn default() -> Self {
```

Comment:
Why do we introduce the default constructor? Seems non-intuitive to transition from 0 count to a non-zero count